File: executor/executor/qemu.py
from executor.http_server import CredentialServer
from .github import GitHubRunnerStatusWatcher
from .qmp import QMPClient
from .utils import log, Timer
from dataclasses import dataclass, field
from pathlib import Path
from typing import List, Optional
import os
import pathlib
import shutil
import subprocess
import tempfile


# How many seconds to wait after a graceful shutdown signal before killing the
# virtual machine.
GRACEFUL_SHUTDOWN_TIMEOUT = 60

# Architecture-specific QEMU flags and BIOS blob URL.
QEMU_ARCH = {
    "x86_64": {
        "bios": None,
        "cpu_model": None,
        # Standard x86_64 machine with hardware acceleration.
        "machine": "pc,accel=kvm",
    },
    "aarch64": {
        # Installed with `sudo apt-get install qemu-efi-aarch64`
        "bios": "/usr/share/qemu-efi-aarch64/QEMU_EFI.fd",
        # Use the host's CPU variant.
        "cpu_model": "host",
        # Virtual AArch64 machine with hardware acceleration.
        "machine": "virt,gic_version=3,accel=kvm",
    },
}


class VM:
    def __init__(self, cli, instance, image, runner):
        self._cli = cli
        self._base = image
        self._vm_timeout = instance["timeout-seconds"]
        self._cpu = instance["cpu-cores"]
        self._ram = instance["ram"]
        self._disk = instance["root-disk"]
        self._runner = runner

        # Once the GitHub Actions build start, the VM won't shutdown anymore when requested by the
        # outside world (for example due to a SIGTERM, or a new image being available), as that
        # would kill the CI build running in the VM.
        self._prevent_external_shutdowns = False

        self._arch = instance["arch"]
        if self._arch not in QEMU_ARCH:
            raise RuntimeError(f"unsupported architecture: {self._arch}")

        self._path = pathlib.Path(tempfile.mkdtemp())
        self._path_root = self._path / "root.qcow2"

        self._process = None
        self._qmp_shutdown_path = self._path / "shutdown.sock"

        self._copy_base_image()

    def _copy_base_image(self):
        if self._path.exists():
            shutil.rmtree(self._path)

        self._path.mkdir(exist_ok=True)

        log("creating the disk image")
        subprocess.run(
            [
                "qemu-img",
                "create",
                # Path of the base image.
                "-b",
                str(pathlib.Path(self._base).resolve()),
                # Use a Copy on Write filesystem, to avoid having to copy the whole
                # base image every time we start a VM.
                "-f",
                "qcow2",
                # Explicitly set format of backing file
                "-F",
                "qcow2",
                # Path of the destination image.
                str(self._path_root.resolve()),
                # New size of the disk.
                self._disk,
            ],
            stdout=subprocess.DEVNULL,
            check=True,
        )

    def run(self, gh):
        if self._process is not None:
            raise RuntimeError("this VM was already started")

        qemu = QemuInvocation(
            cpu_cores=self._cpu,
            memory=self._ram,
            drive=f"file={self._path_root},media=disk,if=virtio",
            bios=QEMU_ARCH[self._arch]["bios"],
            cpu_model=QEMU_ARCH[self._arch]["cpu_model"],
            machine=QEMU_ARCH[self._arch]["machine"],
            qemu_binary=f"qemu-system-{self._arch}",
        )

        # This QMP port is used by the shutdown() method to send the
        # shutdown signal to the QEMU VM instead of killing it.
        qemu.qmp_sockets.append(self._qmp_shutdown_path)

        if self._cli.ssh_port is not None:
            # We only bind to SSH when a port is requested.
            qemu.net_user.append(f"hostfwd=tcp:127.0.0.1:{self._cli.ssh_port}-:22")

        # Pass the credential asking the runner not to shutdown. This is the first credential we add
        # because it has to be passed to the VM even if the following credentials get truncated or
        # similar (as this credential is used for debugging).
        if self._cli.no_shutdown_after_job:
            qemu.smbios_11.append("value=io.systemd.credential:gha-inhibit-shutdown=1")

        jitconfig = CredentialServer("gha-jitconfig-url", self._runner.jitconfig)
        jitconfig.configure_qemu(qemu)

        log("starting the virtual machine")
        self._process = qemu.spawn()

        if self._cli.ssh_port is not None:
            print()
            print("You can connect to the VM with SSH:")
            print()
            print(
                "    "
                "ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null "
                f"-p {self._cli.ssh_port} manage@127.0.0.1"
            )
            print()

        GitHubRunnerStatusWatcher(gh, self._runner.id, self._gha_build_started).start()

        try:
            self._process.wait()
        except KeyboardInterrupt:
            self._shutdown()

        # Shutdown signal was successful, wait for clean shutdown
        try:
            if self._process is not None:
                self._process.wait()
        except KeyboardInterrupt:
            self._kill()

    def request_shutdown(self, reason):
        if self._prevent_external_shutdowns:
            log(f"did not shutdown due to {reason} because a build is running")
        else:
            log(f"shutting down the VM due to {reason}")
            self._shutdown()

    def _shutdown(self):
        if self._process is None:
            raise RuntimeError("can't shutdown a stopped VM")

        # QEMU allows interacting with the VM through the "monitoring port",
        # using Telnet as the protocol. This tries to connect to the monitoring
        # port to send the graceful shutdown signal. If it fails, we're forced
        # to hard-kill the virtual machine.
        try:
            qmp = QMPClient(self._qmp_shutdown_path)
            qmp.shutdown_vm()
        except Exception as e:
            print("failed to gracefully shutdown the VM:", e)
            self._kill()
            return

        log("sent shutdown signal to the VM")

        Timer(
            "graceful-shutdown-timeout", self._kill, GRACEFUL_SHUTDOWN_TIMEOUT
        ).start()

    def _kill(self):
        if self._process is None:
            raise RuntimeError("can't kill a stopped VM")

        self._process.kill()
        self._process = None

        log("killed the virtual machine")

    def cleanup(self):
        shutil.rmtree(str(self._path))

    def _gha_build_started(self):
        self._prevent_external_shutdowns = True
        Timer("vm-timeout", self._shutdown, self._vm_timeout).start()


@dataclass
class QemuInvocation:
    bios: Optional[str]
    cpu_cores: int
    cpu_model: Optional[str]
    drive: str
    machine: str
    memory: int
    qemu_binary: str

    qmp_sockets: List[Path] = field(default_factory=list)
    net_user: List[str] = field(default_factory=list)
    smbios_11: List[str] = field(default_factory=list)

    def spawn(self) -> subprocess.Popen:
        def preexec_fn():
            # Don't forward signals to QEMU
            os.setpgrp()

        cmd = [
            self.qemu_binary,
            # Machine to emulate.
            "-machine",
            self.machine,
            # Reserved RAM for the virtual machine.
            "-m",
            str(self.memory),
            # Allocated cores for the virtual machine.
            "-smp",
            str(self.cpu_cores),
            # Prevent QEMU from showing a graphical console window.
            "-display",
            "none",
            # Mount the VM image as the root drive.
            "-drive",
            self.drive,
            # Enable networking inside the VM.
            "-net",
            "nic,model=virtio",
            # Port forwarding configuration.
            "-net",
            "user" + "".join(f",{param}" for param in self.net_user),
        ]

        if self.cpu_model is not None:
            cmd += ["-cpu", self.cpu_model]

        if self.bios is not None:
            cmd += ["-bios", self.bios]

        for socket in self.qmp_sockets:
            cmd += ["-qmp", f"unix:{socket},server,nowait"]

        for param in self.smbios_11:
            cmd += ["-smbios", f"type=11,{param}"]

        return subprocess.Popen(cmd, preexec_fn=preexec_fn)

File: executor/src/qemu.rs
use crate::credential_server::CredentialServer;
use crate::github::{CreatedRunner, GitHub};
use crate::qmp::QmpClient;
use crate::{InstanceArch, InstanceSpec, Size};
use anyhow::{Error, bail};
use std::ffi::{OsStr, OsString};
use std::os::unix::process::CommandExt;
use std::path::{Path, PathBuf};
use std::pin::Pin;
use std::process::{Command, Stdio};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, Mutex};
use std::time::Duration;
use tempfile::NamedTempFile;
use tokio::process::Command as TokioCommand;
use tokio::signal::unix::{SignalKind, signal};
use tokio::sync::mpsc;
use tokio::time::sleep;

pub async fn start_vm(
    github: &GitHub,
    spec: &InstanceSpec,
    image: &Path,
    runner: &CreatedRunner,
    shutdown_request: ShutdownRequester,
    opts: VmOptions,
) -> Result<(), Error> {
    let root = prepare_root_filesystem(spec, image)?;

    let mut qemu = QemuInvocation {
        arch: spec.arch,
        cpu_cores: spec.cpu_cores,
        drive_path: root.path().into(),
        memory: spec.ram,
        qmp_sockets: Vec::new(),
        net_user: Vec::new(),
        smbios_11: Vec::new(),
    };

    // QMP lets us control QEMU. We need this to gracefully shutdown the VM.
    let qmp_sock = NamedTempFile::new()?;
    qemu.qmp_sockets.push(qmp_sock.path().into());

    if let Some(port) = opts.ssh_port {
        // We only bind to SSH when a port is requested.
        qemu.net_user
            .push(format!("hostfwd=tcp:127.0.0.1:{port}-:22"));
    }

    // Pass the credential asking the runner not to shutdown. This is the first credential we add
    // because it has to be passed to the VM even if the following credentials get truncated or
    // similar (as this credential is used for debugging).
    if !opts.shutdown_after_job {
        qemu.smbios_11.push(Smbios11::Value(
            "io.systemd.credential:gha-inhibit-shutdown=1".into(),
        ));
    }

    let jitconfig = CredentialServer::new("gha-jitconfig-url", &runner.jitconfig).await?;
    let _guard = jitconfig.configure_qemu(&mut qemu)?;

    eprintln!("starting the VM");
    let mut child = TokioCommand::from(qemu.into_command()).spawn()?;

    if let Some(port) = opts.ssh_port {
        eprintln!();
        eprintln!("you can now connect to the VM with SSH:");
        eprintln!();
        eprintln!(
            "    ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -p {} \
             manage@127.0.0.1",
            port
        );
        eprintln!();
    }

    // When a SIGTERM is received, request a graceful shutdown instead of killing the process. This
    // allows us to reboot the host machine without killing any incoming job, as shutdown requests
    // are ignored when a job is running.
    {
        let shutdown_request = shutdown_request.clone();
        let mut sigterm = signal(SignalKind::terminate())?;
        tokio::spawn(async move {
            loop {
                sigterm.recv().await;
                shutdown_request.request_shutdown(ShutdownReason::SigTerm);
            }
        });
    }

    // Inhibit shutdown requests once a job starts.
    {
        let runner_id = runner.id;
        let github = github.clone();
        let shutdown_request = shutdown_request.clone();
        tokio::spawn(async move {
            github
                .run_after_build_started(runner_id, move || {
                    shutdown_request
                        .inner
                        .allowed
                        .store(false, Ordering::Relaxed);
                })
                .await;
        });
    }

    // When a graceful shutdown is started, this future will be replaced from a future that
    // never returns to a sleep. The loop below will kill the VM when this future resolves.
    let mut graceful_shutdown_timeout: Pin<Box<dyn Future<Output = ()>>> =
        Box::pin(std::future::pending());
    let mut graceful_shutdown_started = false;

    let mut shutdown_request = shutdown_request
        .inner
        .receiver
        .lock()
        .unwrap()
        .take()
        .expect("another VM already took the receiver");

    let mut first_ctrlc = true;
    loop {
        // Lint is #[expect]ed because there might be more reasons to do multiple iterations of
        // this loop than just graceful shutdowns.
        #[expect(unused_assignments)]
        let mut do_graceful_shutdown = false;

        tokio::select! {
            _ = child.wait() => return Ok::<_, Error>(()),

            reason = shutdown_request.recv() => {
                eprintln!("graceful shutting down due to {}", reason.unwrap().description());
                do_graceful_shutdown = true;
            }

            _ = &mut graceful_shutdown_timeout => {
                eprintln!("graceful shutdown timeout reached, killing the VM");
                child.kill().await?;
                return Ok(());
            }

            _ = tokio::signal::ctrl_c() => {
                if first_ctrlc {
                    first_ctrlc = false;

                    eprintln!("pressed Ctrl+C, gracefully shutting down (press again to kill)");
                    do_graceful_shutdown = true;
                } else {
                    eprintln!("pressed Ctrl+C again, killing the VM");
                    child.kill().await?;
                    return Ok(());
                }
            }
        }

        if do_graceful_shutdown && !graceful_shutdown_started {
            match send_graceful_shutdown(&qmp_sock.path()).await {
                Ok(()) => {
                    graceful_shutdown_started = true;
                    graceful_shutdown_timeout = Box::pin(sleep(Duration::from_secs(60)));
                }
                Err(err) => {
                    eprintln!("graceful shutdown failed: {err}, killing the VM");
                    child.kill().await?;
                    return Ok(());
                }
            }
        }
    }
}

async fn send_graceful_shutdown(qmp_socket: &Path) -> Result<(), Error> {
    let mut qmp = QmpClient::new(qmp_socket).await?;
    qmp.shutdown_vm().await?;
    Ok(())
}

pub struct VmOptions {
    pub ssh_port: Option<u16>,
    pub shutdown_after_job: bool,
}

fn prepare_root_filesystem(spec: &InstanceSpec, image: &Path) -> Result<NamedTempFile, Error> {
    let dest = NamedTempFile::new()?;

    eprintln!("creating the root filesystem image");
    let status = Command::new("qemu-img")
        .arg("create")
        // Path of the base image.
        .args([OsStr::new("-b"), image.as_os_str()])
        // Use a Copy on Write filesystem, to avoid having to copy the whole base image every time
        // we start the VM.
        .args(["-f", "qcow2"])
        // Explicitly set the format of the backing image file.
        .args(["-F", "qcow2"])
        // Path to the destination image.
        .arg(dest.path())
        // Size of the disk we are creating.
        .arg(spec.root_disk.to_string())
        // We don't care about the output of the command (we do about errors on stderr).
        .stdout(Stdio::null())
        .status()?;
    if !status.success() {
        bail!("attempting to create the root disk exited with {status}");
    }

    Ok(dest)
}

#[derive(Clone)]
pub struct ShutdownRequester {
    inner: Arc<ShutdownRequesterInner>,
}

impl ShutdownRequester {
    pub fn new() -> Self {
        let (sender, receiver) = mpsc::unbounded_channel();
        Self {
            inner: Arc::new(ShutdownRequesterInner {
                allowed: AtomicBool::new(true),
                sender,
                receiver: Mutex::new(Some(receiver)),
            }),
        }
    }

    pub fn request_shutdown(&self, reason: ShutdownReason) {
        if !self.inner.allowed.load(Ordering::Relaxed) {
            eprintln!(
                "shutdown requested due to {}, but a job is currently running, skipping it",
                reason.description()
            );
            return;
        }
        self.inner
            .sender
            .send(reason)
            .expect("failed to send the shutdown request");
    }
}

struct ShutdownRequesterInner {
    allowed: AtomicBool,
    sender: mpsc::UnboundedSender<ShutdownReason>,
    receiver: Mutex<Option<mpsc::UnboundedReceiver<ShutdownReason>>>,
}

pub enum ShutdownReason {
    NewImages,
    SigTerm,
}

impl ShutdownReason {
    fn description(&self) -> &str {
        match self {
            ShutdownReason::NewImages => "new images being available",
            ShutdownReason::SigTerm => "SIGTERM signal",
        }
    }
}

pub(crate) struct QemuInvocation {
    pub(crate) arch: InstanceArch,
    pub(crate) cpu_cores: u32,
    pub(crate) drive_path: PathBuf,
    pub(crate) memory: Size,

    pub(crate) qmp_sockets: Vec<PathBuf>,
    pub(crate) net_user: Vec<String>,
    pub(crate) smbios_11: Vec<Smbios11>,
}

impl QemuInvocation {
    fn into_command(self) -> Command {
        let mut cmd = Command::new(match self.arch {
            InstanceArch::X86_64 => "qemu-system-x86_64",
            InstanceArch::Aarch64 => "qemu-system-aarch64",
        });
        cmd
            // Machine to emulate
            .args([
                "-machine",
                match self.arch {
                    InstanceArch::X86_64 => "pc,accel=kvm",
                    InstanceArch::Aarch64 => "virt,gic_version=3,accel=kvm",
                },
            ])
            // Allocated RAM
            .args(["-m", &self.memory.to_string()])
            // Allocated virtual cores
            .args(["-smp", &self.cpu_cores.to_string()])
            // Prevent QEMU from showing a graphical console window.
            .args(["-display", "none"])
            // Mount the VM image inside the VM
            .args([
                concat_os_string(&[&"-drive"]),
                concat_os_string(&[&"file=", &self.drive_path, &",media=disk,if=virtio"]),
            ])
            // Enable networking
            .args(["-net", "nic,model=virtio"])
            // Port forwarding configuration
            .args([
                "-net".to_string(),
                format!(
                    "user{}",
                    self.net_user
                        .iter()
                        .map(|p| format!(",{p}"))
                        .collect::<String>()
                ),
            ]);

        match self.arch {
            InstanceArch::X86_64 => {}
            InstanceArch::Aarch64 => {
                cmd.args(["-cpu", "host"])
                    .args(&["-bios", "/usr/share/qemu-efi-aarch64/QEMU_EFI.fd"]);
            }
        }

        for socket in &self.qmp_sockets {
            cmd.arg("-qmp")
                .arg(concat_os_string(&[&"unix:", socket, &",server,nowait"]));
        }
        for param in &self.smbios_11 {
            match param {
                Smbios11::Path(path) => cmd
                    .arg("-smbios")
                    .arg(concat_os_string(&[&"type=11,path=", path])),
                Smbios11::Value(value) => cmd.arg("-smbios").arg(format!("type=11,value={value}")),
            };
        }

        // Don't propagate signals into the VM.
        cmd.process_group(0);

        cmd
    }
}

pub(crate) enum Smbios11 {
    Path(PathBuf),
    Value(String),
}

fn concat_os_string(components: &[&dyn AsRef<OsStr>]) -> OsString {
    let mut result = OsString::new();
    for component in components {
        result.push(component);
    }
    result
}
